src/subset/pretty_print.rs

```diff
@@ -47,18 +49,63 @@ impl Ord for ParenCtx {
         match (self, other) {
             (P::Not, _) => Ordering::Greater,
 
-            (P::Op, P::Not) => Ordering::Less,
-            (P::Op, _) => Ordering::Greater,
+            (P::Index, P::Not) => Ordering::Less,
+            (P::Index, _) => Ordering::Greater,
+
+            (P::Mul, P::Not) | (P::Mul, P::Index) => Ordering::Less,
```

Comment:
Since this code was written, rust started supporting nested patterns so you can write something like:
```
(P::Mul, P::Not | P::Index)
```
Which might be easier to read